auth/clientcredentials/clientcredentials.go

```diff
@@ -0,0 +1,229 @@
+	var span trace.Span
+	ctx, span = tracer.Start(ctx, "clientcredentials.requireScope")
+	defer func() {
+		if err != nil {
+			span.RecordError(err)
+			span.SetStatus(otelcodes.Error, "check failed")
+		}
+		span.End()
+	}()
+
+	token, err := extractBearerContents(headers)
+	if err != nil {
+		return nil, connect.NewError(connect.CodeUnauthenticated,
+			errors.Wrap(err, "invalid authorization header"))
+	}
+
+	result, err := i.introspector.IntrospectToken(ctx, token)
+	if err != nil {
+		return nil, internalError(ctx, i.logger, err, "unable to validate token")
+	}
+	span.SetAttributes(attribute.String("client_id", result.ClientID))
+	info := &ClientInfo{ClientID: result.ClientID, TokenExpiresAt: result.ExpiresAt, TokenScopes: result.Scopes}
+	logger := i.logger.With(info.LogFields()...)
+
+	// Active encapsulates whether the token is active, including expiration.
+	if !result.Active {
+		// Record detailed error in span, and return an opaque one
+		span.SetAttributes(attribute.String("full_error", "inactive token"))
+		return info, connect.NewError(connect.CodePermissionDenied, errors.New("permission denied"))
+	}
+
+	// Check for our required scope.
+	for _, required := range requiredScopes {
+		if !result.Scopes.Match(required) {
+			// Record detailed error in span and logs
+			err = errors.Newf("got scopes %+v, required: %+v", result.Scopes, requiredScopes)
+			span.SetAttributes(attribute.String("full_error", err.Error()))
+			logger.Error("attempt to authenticate using SAMS token without required scope",
+				log.Error(err))
+			// Return an opaque error
+			return info, connect.NewError(connect.CodePermissionDenied, errors.New("insufficient scope"))
+		}
+	}
+
+	return info, nil
+}
+
+func extractSchemaRequiredScopes(spec connect.Spec, extension *protoimpl.ExtensionInfo) (scopes.Scopes, error) {
+	method, ok := spec.Schema.(protoreflect.MethodDescriptor)
+	if !ok {
+		return nil, errors.Newf("expected protoreflect.MethodDescriptor, got %T", spec.Schema)
+	}
+
+	value := method.Options().ProtoReflect().Get(extension.TypeDescriptor())
+	if !value.IsValid() {
+		return nil, errors.Newf("extension field %s not valid", extension.TypeDescriptor().FullName())
+	}
+	list := value.List()
+	if list.Len() == 0 {
+		return nil, errors.Newf("extension field %s cannot be empty", extension.TypeDescriptor().FullName())
+	}
+
+	requiredScopes := make(scopes.Scopes, list.Len())
+	for i := 0; i < list.Len(); i++ {
+		requiredScopes[i] = scopes.Scope(list.Get(i).String())
+	}
+	return requiredScopes, nil
+}
+
+func extractBearerContents(h http.Header) (string, error) {
+	authHeader := h.Get("Authorization")
+	if authHeader == "" {
+		return "", errors.New("no token provided in Authorization header")
+	}
+	typ := strings.SplitN(authHeader, " ", 2)
+	if len(typ) != 2 {
+		return "", errors.New("token type missing in Authorization header")
+	}
+	if strings.ToLower(typ[0]) != "bearer" {
```

Comment:
```suggestion
	if strings.EqualFold(typ[0], "bearer") {
```

can avoid string allocation.